### DISCUSSION TITLE
Without the index built on scalar fields, Milvus will perform a brute-force search by default?

### DISCUSSION THREAD

**Discussion by LiJie20190102** [milvus-io/milvus#23407] [Q&A and General discussion] [answered]
[Most of the vector index types supported by Milvus use approximate nearest neighbors search (ANNS), including:](https://milvus.io/docs/build_index.md#Build-an-Index-on-Vectors)

Without the index built on vectors, Milvus will perform a brute-force search by default.But without the index built on scalar fields, Milvus will perform a brute-force search by default?

**Comment by yhmo** [comment]
Index for a scalar field is optional, you can build index for scalar field: https://milvus.io/docs/build_scalar_index.md
For numeric scalar field, the index is a simple sorted array.
For varchar field, the index is a tree-based index.

Building an index for a scalar field can slightly improve the search performance if you perform a "filtering search" on this scalar field.

**Reply by LiJie20190102** [reply]
https://milvus.io/docs/scalar_index.md#Scalar-field-indexing-algorithms
The official website says that all scalar fields will be indexed. I would like to ask if I have three scalars (id: INT64, name: varchar, age: INT64), where id is the primary key. Will the index be automatically created at this time? If it is automatically created, will only the primary key id be created, or will all three scalar fields be created?
At the same time, if I only use scalar fields for search (not vector search), there shouldn't be any problems, right

**Reply by yhmo** [reply]
> https://milvus.io/docs/scalar_index.md#Scalar-field-indexing-algorithms The official website says that all scalar fields will be indexed. I would like to ask if I have three scalars (id: INT64, name: varchar, age: INT64), where id is the primary key. Will the index be automatically created at this time? If it is automatically created, will only the primary key id be created, or will all three scalar fields be created? At the same time, if I only use scalar fields for search (not vector search), there shouldn't be any problems, right

The doc is out of date. Milvus only automatically build index for the primary key field. Users need to manually build index for other scalar fields. We will update the doc later.

To build index for scalar fields, refer this doc: https://milvus.io/docs/build_scalar_index.md

Search can work well even if you didn't build index for the scalar fields.

**Reply by LiJie20190102** [reply]
I created a collection using the Java SDK and specified a primary key, but did not set an index for the primary key.

![image](https://user-images.githubusercontent.com/53458004/231758426-dd63b1bd-2d8a-4918-a42a-56ece4646696.png)

**Reply by yhmo** [reply]
It automatically build index for the primary key.

**Reply by LiJie20190102** [reply]
Okay, that means it won't show, right? Thank you

**Reply by yhmo** [reply]
> Okay, that means it won't show, right? Thank you

Right

**Reply by ducanh997** [reply]
Does it still automatically create an index on a varchar PK field in the 2.3.3 version? Thank you.

**Reply by yhmo** [reply]
> Does it still automatically create an index on a varchar PK field on the 2.3.3 version? Thank you.

Yes, it automatically creates a sorted-index on a varchar PK field. If you use create_index() to specify index to the varchar PK, the sorted-index will be rebuilt to Trie tree index.

**Reply by ducanh997** [reply]
Thank @yhmo. What will happen if I create_index() on a varchar PK and then drop it? Will the trie index be rebuilt to the sorted index.

**Reply by yhmo** [reply]
create_index() creates trie index, if you use drop_index(), the trie index is dropped, but the sorted index is still there.

**Comment by LiJie20190102** [comment]
Is using the Java SDK like this? But indexType and metricType are required here.
![image](https://user-images.githubusercontent.com/53458004/231936767-3c494dab-9a5e-49fd-8c6a-06cd9d6b609f.png)

![image](https://user-images.githubusercontent.com/53458004/231937032-f8ad880d-5102-4bea-a13c-0e8c822bdfcd.png)
![image](https://user-images.githubusercontent.com/53458004/231937105-38ff5e6d-82a0-4ce6-a218-97b1a27107a6.png)

What type of indexType and metricType should I specify.

**Reply by yhmo** [reply]
Java SDK example: https://github.com/milvus-io/milvus-sdk-java/blob/66d88d269b7cdce3d78e3bbd80abc4a1d3e8bd45/examples/main/java/io/milvus/GeneralExample.java#L252

**Reply by yhmo** [reply]
Example to create a HNSW index:
```
CreateIndexParam indexParam = CreateIndexParam.newBuilder()
                .withCollectionName(collectionName)
                .withFieldName(vectorFieldName)
                .withIndexName("xxx")
                .withIndexType(IndexType.HNSW)
                .withMetricType(MetricType.L2)
                .withExtraParam("{\"M\":16,\"efConstruction\":64}")
                .build();

R<RpcStatus> createIndexR = client.createIndex(indexParam);
```

**Reply by LiJie20190102** [reply]
https://milvus.io/docs/build_scalar_index.md#Build-index
1、The official website says that for scalar fields, there is no need for additional index parameters, and only the field needs to be provided_ Name and index_ name is enough? Is the official website outdated?
2、For scalar fields, if they are not primary keys, they need to be manually indexed. According to the Java SDK, I need to provide an index type. So what is the range of index types for scalar fields?

**Reply by LiJie20190102** [reply]
> https://milvus.io/docs/build_scalar_index.md#Build-index 1、The official website says that for scalar fields, there is no need for additional index parameters, and only the field needs to be provided_ Name and index_ name is enough? Is the official website outdated? 2、For scalar fields, if they are not primary keys, they need to be manually indexed. According to the Java SDK, I need to provide an index type. So what is the range of index types for scalar fields?

对应中文意思：

对于官网：https://milvus.io/docs/build_scalar_index.md#Build-index
1、官网上面说要在标量字段上建立索引，不需要设置任何索引参数。是不是只需要提供字段名称和索引类型就可以了
2、对于java SDK 来说，创建标量非主键字段时，也是使用CreateIndexParam来创建么，好像必须要提供IndexType和MetricType。是不是标量非主键字段来说，有别的创建方式。

麻烦了，帮忙看下。

**Reply by yhmo** [reply]
https://milvus.io/docs/build_scalar_index.md
标量建索引只需要指定列名和索引名就可以。如果用java sdk，就像这样：
```
CreateIndexParam indexParam = CreateIndexParam.newBuilder()
                .withCollectionName(collectionName)
                .withFieldName(fieldName)
                .withIndexName("xxx")
                .build();

R<RpcStatus> createIndexR = client.createIndex(indexParam);
```

**Reply by LiJie20190102** [reply]
我像你那样使用java SDK创建标量非主键字段的索引时，会报错：
![image](https://user-images.githubusercontent.com/53458004/231990432-30c5ea2d-3992-40d5-8ab3-1fd881d4c70a.png)
源码：
![image](https://user-images.githubusercontent.com/53458004/231990293-e24dbbbe-0a4e-4c5c-843b-5ec1ad5d991d.png)

**Reply by yhmo** [reply, marked as answer]
数值型的标量就没必要建索引了，你可以给VARCHAR标量建索引：
```
CreateIndexParam indexParam = CreateIndexParam.newBuilder()
                .withCollectionName(collectionName)
                .withFieldName(fieldName)
                .withIndexType(IndexType.TRIE)
                .withIndexName("xxx")
                .build();

R<RpcStatus> createIndexR = client.createIndex(indexParam);
```

**Reply by LiJie20190102** [reply]
这种方式是好使的，谢谢了
